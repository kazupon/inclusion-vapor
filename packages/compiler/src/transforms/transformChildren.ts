// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformChildren.ts

import { IRNodeTypes, DynamicFlag } from '../ir'
import { transformNode } from '../transform'

import type { NodeTransform } from './types'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT ||
    node.type === 'Fragment' ||
    node.type === 'Element' ||
    node.type === 'InlineComponent'
  if (!isFragment) {
    return
  }

  const children = node.children || []
  for (const [i, child] of children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }
}
