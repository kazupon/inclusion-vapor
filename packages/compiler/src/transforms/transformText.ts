// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformText.ts

import { createSimpleExpression } from '@vue-vapor/compiler-dom'
import { isString } from '@vue-vapor/shared'
import { parseExpression } from '@babel/parser'
import {
  DynamicFlag,
  IRNodeTypes,
  isSvelteElement,
  isSvelteText,
  isSvelteMustacheTag,
  convertToSourceLocation
} from '../ir'
import { isConstantExpression, getLiteralExpressionValue } from './utils'

import type { SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type { ParseResult as BabelParseResult } from '@babel/parser'
import type { Expression as BabelExpression } from '@babel/types'
import type { TransformContext } from './context'
import type { NodeTransform } from './types'
import type {
  SvelteTemplateNode,
  RootNode,
  SvelteElement,
  SvelteText,
  SvelteMustacheTag
} from '../ir'

type TextLike = SvelteText | SvelteMustacheTag

const seen = new WeakMap<TransformContext<RootNode>, WeakSet<SvelteTemplateNode | RootNode>>()

export const transformText: NodeTransform = (node, context) => {
  if (!seen.has(context.root)) {
    seen.set(context.root, new WeakSet())
  }

  if (seen.get(context.root)!.has(node)) {
    context.dynamic.flags |= DynamicFlag.NON_TEMPLATE
    return
  }

  const children = node.children || []
  if (isSvelteElement(node) && isAllTextLike(children)) {
    // TODO: should more tested for component
    processTextLikeContainer(children, context as TransformContext<SvelteElement>)
  } else if (isSvelteMustacheTag(node)) {
    // TODO: should more tested for MustacheTag expression
    // e.g. ConditionalExpression, LogicalExpression and CallExpression ...
    processTextLike(context as TransformContext<SvelteMustacheTag>)
  } else if (isSvelteText(node)) {
    context.template += node.data
  }
}

function processTextLike(context: TransformContext<SvelteMustacheTag>): void {
  const children = context.parent!.node.children || []
  const nexts = children.slice(context.index)
  const idx = nexts.findIndex(n => !isTextLike(n))
  const nodes = (idx > -1 ? nexts.slice(0, idx) : nexts) as Array<TextLike>
  const id = context.reference()
  const values = nodes.map(node => createTextLikeExpression(node, context))

  context.dynamic.flags |= DynamicFlag.INSERT | DynamicFlag.NON_TEMPLATE
  context.registerOperation({
    type: IRNodeTypes.CREATE_TEXT_NODE,
    id,
    values,
    // eslint-disable-next-line unicorn/no-array-callback-reference
    effect: !values.every(isConstantExpression) && !context.inVOnce
  })
}

function processTextLikeContainer(
  children: TextLike[],
  context: TransformContext<SvelteElement>
): void {
  const values = children.map(child => createTextLikeExpression(child, context))
  // eslint-disable-next-line unicorn/no-array-callback-reference
  const literals = values.map(getLiteralExpressionValue)
  if (literals.every(l => l != undefined)) {
    context.childrenTemplate = literals.map(String)
  } else {
    context.registerEffect(values, {
      type: IRNodeTypes.SET_TEXT,
      element: context.reference(),
      values
    })
  }
}

function createTextLikeExpression(node: TextLike, context: TransformContext): SimpleExpressionNode {
  seen.get(context.root)!.add(node)
  // prettier-ignore
  return isSvelteText(node)
    ? createSimpleExpression(node.data, true, convertToSourceLocation(node, (node as unknown as { raw: string }).raw || node.data))
    : resolveSimpleExpression(node, context as TransformContext<SvelteMustacheTag>)
}

function resolveSimpleExpression<T extends SvelteMustacheTag>(
  node: SvelteMustacheTag,
  context: TransformContext<T>
): SimpleExpressionNode {
  const { expression } = node

  const content =
    expression.type === 'Identifier'
      ? expression.name
      : (expression.type === 'Literal'
        ? expression.raw || ''
        : context.ir.source.slice(node.start, node.end))
  const loc = expression.loc || convertToSourceLocation(node, content) // FIXME: twaeak loc type

  let ast: false | BabelParseResult<BabelExpression> = false
  const isStatic =
    expression.type === 'Identifier'
      ? false
      : expression.type === 'Literal' && !isString(expression.value)
  if (!isStatic && context.options.prefixIdentifiers) {
    ast = parseExpression(`${content}`, {
      sourceType: 'module',
      plugins: context.options.expressionPlugins
    })
  }

  const exp = createSimpleExpression(
    content,
    isStatic,
    loc as ReturnType<typeof convertToSourceLocation>
  )
  exp.ast = ast ?? null // eslint-disable-line unicorn/no-null
  return exp
}

function isAllTextLike(children: SvelteTemplateNode[]): children is TextLike[] {
  return (
    children.length > 0 &&
    // eslint-disable-next-line unicorn/no-array-callback-reference
    children.every(isTextLike) &&
    // at least one an mustache tag
    children.some(n => n.type === 'MustacheTag')
  )
}

function isTextLike(node: SvelteTemplateNode): node is TextLike {
  // TODO: should more tested for MustacheTag expression
  // e.g. ConditionalExpression, LogicalExpression and CallExpression ...
  return isSvelteText(node) || isSvelteMustacheTag(node)
}
