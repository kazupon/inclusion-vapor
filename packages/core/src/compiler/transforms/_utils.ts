import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
