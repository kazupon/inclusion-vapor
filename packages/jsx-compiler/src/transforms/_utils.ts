import { parse } from '@babel/parser'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, JSXElement, JSXFragment } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const {
      body: [statement]
    } = parse(source, {
      sourceType: 'module',
      plugins: ['jsx']
    }).program
    let children!: JSXElement[] | JSXFragment['children']
    if (statement.type === 'ExpressionStatement') {
      children =
        statement.expression.type === 'JSXFragment'
          ? statement.expression.children
          : (statement.expression.type === 'JSXElement'
            ? [statement.expression]
            : [])
    }

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children,
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
