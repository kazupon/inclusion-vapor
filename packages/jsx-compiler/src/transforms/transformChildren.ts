// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `unplugin-vue-jsx-vapor`
// Author: zhiyuanzmj (https://github.com/zhiyuanzmj) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/unplugin/unplugin-vue-jsx-vapor
// Code url: https://github.com/unplugin/unplugin-vue-jsx-vapor/tree/main/src/core/compiler

import { isComponentNode } from './utils'
import { IRNodeTypes, DynamicFlag } from '../ir'
import { transformNode } from '../transform'

import type { BabelNode, IRDynamicInfo } from '../ir'
import type { NodeTransform } from './types'
import type { TransformContext } from './context'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT || node.type === 'JSXFragment' || isComponentNode(node)

  if (node.type !== 'JSXElement' && !isFragment) {
    return
  }

  ;[...node.children].forEach((child, index) => {
    if (child.type === 'JSXText' && !child.value.trim()) {
      child.value = ' '
      if (!index) {
        node.children.splice(0, 1)
      } else if (index === node.children.length) {
        node.children.splice(-1, 1)
      }
    }
  })

  for (const [i, child] of node.children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<BabelNode>)
  }
}

function processDynamicChildren(context: TransformContext<BabelNode>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length > 0) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`

          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())

          context.registerOperation({
            type: IRNodeTypes.INSERT_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference(),
            anchor
          })
        } else {
          context.registerOperation({
            type: IRNodeTypes.PREPEND_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference()
          })
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length > 0) {
    context.registerOperation({
      type: IRNodeTypes.INSERT_NODE,
      elements: prevDynamics.map(child => child.id!),
      parent: context.reference()
    })
  }
}
