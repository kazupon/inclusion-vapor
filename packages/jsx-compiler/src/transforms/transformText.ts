// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `unplugin-vue-jsx-vapor`
// Author: zhiyuanzmj (https://github.com/zhiyuanzmj) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/unplugin/unplugin-vue-jsx-vapor
// Code url: https://github.com/unplugin/unplugin-vue-jsx-vapor/tree/main/src/core/compiler

import { DynamicFlag, IRNodeTypes } from '../ir/index.ts'
// import { processConditionalExpression, processLogicalExpression } from './vIf'
// import { processMapCallExpression } from './vFor'
import {
  getLiteralExpressionValue,
  isComponentNode,
  isConstantExpression,
  isMapCallExpression,
  resolveExpression
} from './utils.ts'

import type {
  BabelCallExpression,
  BabelNode,
  BlockIRNode,
  JSXElement,
  JSXExpressionContainer,
  JSXText,
  RootNode
} from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

type TextLike = JSXText | JSXExpressionContainer

const seen = new WeakMap<
  TransformContext<RootNode>,
  WeakSet<TextLike | BlockIRNode['node'] | RootNode>
>()

export const transformText: NodeTransform = (node, context) => {
  if (!seen.has(context.root)) {
    seen.set(context.root, new WeakSet())
  }

  if (seen.get(context.root)!.has(node)) {
    context.dynamic.flags |= DynamicFlag.NON_TEMPLATE
    return
  }

  if (
    node.type === 'JSXElement' &&
    !isComponentNode(node) &&
    // @ts-expect-error -- FIXME
    isAllTextLike(node.children) // eslint-disable-line @typescript-eslint/no-unsafe-argument
  ) {
    processTextLikeContainer(
      // @ts-expect-error -- FIXME
      node.children, // eslint-disable-line @typescript-eslint/no-unsafe-argument
      context as TransformContext<JSXElement>
    )
  } else if (node.type === 'JSXExpressionContainer') {
    switch (node.expression.type) {
      // case 'ConditionalExpression': {
      //   return processConditionalExpression(node.expression, context)
      // }
      // case 'LogicalExpression': {
      //   return processLogicalExpression(node.expression, context)
      // }
      case 'CallExpression': {
        if (isMapCallExpression(node.expression)) {
          // return processMapCallExpression(node.expression, context)
        } else {
          processCallExpression(node.expression, context)
        }

        break
      }
      default: {
        processTextLike(context as TransformContext<JSXExpressionContainer>)
      }
    }
  } else if (node.type === 'JSXText') {
    context.template += node.value
  }
}

function processTextLike(context: TransformContext<JSXExpressionContainer>) {
  const nexts = context.parent!.node.children?.slice(context.index)
  const idx = nexts.findIndex(n => !isTextLike(n))
  const nodes = (idx > -1 ? nexts.slice(0, idx) : nexts) as Array<TextLike>

  const id = context.reference()
  const values = nodes.map(node => createTextLikeExpression(node, context))

  context.dynamic.flags |= DynamicFlag.INSERT | DynamicFlag.NON_TEMPLATE

  context.registerOperation({
    type: IRNodeTypes.CREATE_TEXT_NODE,
    id,
    values,
    // eslint-disable-next-line unicorn/no-array-callback-reference
    effect: !values.every(isConstantExpression) && !context.inVOnce
  })
}

function processTextLikeContainer(children: TextLike[], context: TransformContext<JSXElement>) {
  const values = children.map(child => createTextLikeExpression(child, context))
  const literals = values.map(getLiteralExpressionValue) // eslint-disable-line unicorn/no-array-callback-reference
  if (literals.every(l => l != undefined)) {
    context.childrenTemplate = literals.map(String)
  } else {
    context.registerEffect(values, {
      type: IRNodeTypes.SET_TEXT,
      element: context.reference(),
      values
    })
  }
}

function createTextLikeExpression(node: TextLike, context: TransformContext) {
  seen.get(context.root)!.add(node)
  return resolveExpression(node, context)
}

function isAllTextLike(children: BabelNode[]): children is TextLike[] {
  return (
    children.length > 0 &&
    // eslint-disable-next-line unicorn/no-array-callback-reference
    children.every(isTextLike) &&
    // at least one an interpolation
    children.some(n => n.type === 'JSXExpressionContainer')
  )
}

function isTextLike(node: BabelNode): node is TextLike {
  return (
    (node.type === 'JSXExpressionContainer' &&
      !(
        node.expression.type === 'ConditionalExpression' ||
        node.expression.type === 'LogicalExpression'
      ) &&
      node.expression.type !== 'CallExpression') ||
    node.type === 'JSXText'
  )
}

function processCallExpression(node: BabelCallExpression, context: TransformContext) {
  context.dynamic.flags |= DynamicFlag.NON_TEMPLATE | DynamicFlag.INSERT
  const root = context.root === context.parent && context.parent.node.children.length === 1
  const tag = `() => ${context.ir.source.slice(node.start!, node.end!)}`

  context.registerOperation({
    type: IRNodeTypes.CREATE_COMPONENT_NODE,
    id: context.reference(),
    tag,
    props: [],
    asset: false,
    root,
    slots: context.slots,
    once: context.inVOnce
  })
}
