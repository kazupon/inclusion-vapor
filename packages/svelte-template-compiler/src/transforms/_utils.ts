import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, RootIRNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (
    source: string,
    overrideOptions: CompilerOptions = {}
  ): {
    ast: RootNode
    ir: RootIRNode
    code: string
    helpers: Set<string>
    vaporHelpers: Set<string>
  } => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
