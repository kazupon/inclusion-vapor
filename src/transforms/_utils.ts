import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type { RootIRNode as VaporRootIRNode } from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode } from '../ir'

export function makeCompile(options: CompilerOptions = {}) {
  return (source: string, overrideOptions: CompilerOptions = {}) => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      prefixIdentifiers: true,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      prefixIdentifiers: true,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
