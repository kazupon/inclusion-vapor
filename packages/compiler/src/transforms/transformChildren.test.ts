import { describe, expect, test } from 'vitest'
import { makeCompile } from './_utils'
import { transformChildren } from './transformChildren'
import { transformElement } from './transformElement'
import { transformText } from './transformText'
import { compile as vaporCompile } from '@vue-vapor/compiler-vapor'

const compile = makeCompile({
  nodeTransforms: [transformText, transformElement, transformChildren]
})

describe('compiler: children transform', () => {
  test.todo('basic')

  test('native elements', () => {
    const source = `<div>
  <div>
    <p>Hello</p>
  </div>
  <p>World</p>
</div>`
    const { ir: _, code, vaporHelpers: __ } = compile(source)
    const expectedResult = vaporCompile(source)
    expect(code).toMatchSnapshot('recieved')
    expect(expectedResult.code).toMatchSnapshot('expected')
    expect(code).toEqual(expectedResult.code)
    // expect(vaporHelpers).toEqual(expectedResult.vaporHelpers)
  })

  test('sibling references', () => {
    const source1 = `<div>
  <p>{ first }</p>
  { second }
  { third }
  <p>{ forth }</p>
</div>`
    const { ir: _, code, vaporHelpers: __ } = compile(source1)
    const source2 = `<div>
  <p>{{ first }}</p>
  {{ second }}
  {{ third }}
  <p>{{ forth }}</p>
</div>`
    const expectedResult = vaporCompile(source2)
    expect(code).toMatchSnapshot('recieved')
    expect(expectedResult.code).toMatchSnapshot('expected')
    expect(code).toEqual(expectedResult.code)
    // expect(vaporHelpers).toEqual(expectedResult.vaporHelpers)
  })

  test('components', () => {
    const source1 = `<div>
  <Foo></Foo>
</div>`
    const { ir: _, code, vaporHelpers: __ } = compile(source1)
    const source2 = `<div>
  <Foo></Foo>
</div>`
    const expectedResult = vaporCompile(source2)
    expect(code).toMatchSnapshot('recieved')
    expect(expectedResult.code).toMatchSnapshot('expected')
    expect(code).toEqual(expectedResult.code)
    // expect(vaporHelpers).toEqual(expectedResult.vaporHelpers)
  })

  test.todo('components has slots with native elements')
  test.todo('components has slots with mustache')
})
