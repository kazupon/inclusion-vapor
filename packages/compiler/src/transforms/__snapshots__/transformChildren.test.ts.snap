// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: children transform > components > expected 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent, insert as _insert, template as _template } from 'vue/vapor';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n1 = t0()
  const n0 = _createComponent(_component_Foo)
  _insert(n0, n1)
  return n1
}"
`;

exports[`compiler: children transform > components > recieved 1`] = `
"import { resolveComponent as _resolveComponent, createComponent as _createComponent, template as _template } from 'vue/vapor';
const t0 = _template("\\n  ")
const t1 = _template("\\n")
const t2 = _template("<div></div>")

export function render(_ctx) {
  const _component_Foo = _resolveComponent("Foo")
  const n3 = t2()
  const n0 = n3.firstChild
  const n0 = t0()
  const n2 = n0.nextSibling
  const n2 = t1()
  const n1 = _createComponent(_component_Foo)
  return [n0, n1, n2, n3]
}"
`;

exports[`compiler: children transform > native elements > expected 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("<div><div><p>Hello</p></div><p>World</p></div>")

export function render(_ctx) {
  const n0 = t0()
  return n0
}"
`;

exports[`compiler: children transform > native elements > recieved 1`] = `
"import { template as _template } from 'vue/vapor';
const t0 = _template("\\n  ")
const t1 = _template("\\n    ")
const t2 = _template("Hello")
const t3 = _template("<p></p>")
const t4 = _template("<div></div>")
const t5 = _template("World")
const t6 = _template("\\n")

export function render(_ctx) {
  const n10 = t4()
  const n0 = n10.firstChild
  const n0 = t0()
  const n5 = n0.nextSibling
  const n5 = t4()
  const n1 = n5.firstChild
  const n1 = t1()
  const n3 = n1.nextSibling
  const n3 = t3()
  const n2 = n3.firstChild
  const n2 = t2()
  const n4 = n3.nextSibling
  const n4 = t0()
  const n6 = n5.nextSibling
  const n6 = t0()
  const n8 = n6.nextSibling
  const n8 = t3()
  const n7 = n8.firstChild
  const n7 = t5()
  const n9 = n8.nextSibling
  const n9 = t6()
  return [n0, n1, n2, n3, n4, n5, n6, n7, n8, n9, n10]
}"
`;

exports[`compiler: children transform > sibling references > expected 1`] = `
"import { next as _next, createTextNode as _createTextNode, insert as _insert, renderEffect as _renderEffect, setText as _setText, template as _template } from 'vue/vapor';
const t0 = _template("<div><p></p> <!><p></p></div>")

export function render(_ctx) {
  const n4 = t0()
  const n0 = n4.firstChild
  const n3 = _next(n0, 2)
  const n2 = n3.nextSibling
  const n1 = _createTextNode(() => [second, " ", third, " "])
  _insert(n1, n4, n3)
  _renderEffect(() => _setText(n0, first))
  _renderEffect(() => _setText(n2, forth))
  return n4
}"
`;

exports[`compiler: children transform > sibling references > recieved 1`] = `
"import { children as _children, next as _next, createTextNode as _createTextNode, renderEffect as _renderEffect, setText as _setText, template as _template } from 'vue/vapor';
const t0 = _template("\\n  ")
const t1 = _template("<p></p>")
const t2 = _template("\\n")
const t3 = _template("<div></div>")

export function render(_ctx) {
  const n11 = t3()
  const n0 = n11.firstChild
  const n0 = t0()
  const n1 = n0.nextSibling
  const n1 = t1()
  const n2 = _children(n1, -1)
  const n3 = n1.nextSibling
  const n3 = t0()
  const n5 = _next(n3, 0)
  const n6 = _next(n5, 0)
  const n7 = _next(n6, 0)
  const n8 = n7.nextSibling
  const n8 = t1()
  const n9 = _children(n8, -1)
  const n10 = n8.nextSibling
  const n10 = t2()
  const n4 = _createTextNode(() => [secon_ctx.d, "\\n  ", thir_ctx.d, "\\n  "])
  _renderEffect(() => _setText(n1, firs_ctx.t))
  _renderEffect(() => _setText(n8, fort_ctx.h))
  return [n0, n1, n3, n4, n8, n10, n11]
}"
`;
